core/src/SYCL/Kokkos_SYCL_Space.cpp

```diff
@@ -261,6 +309,60 @@ SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace, void>::
       "HostSpace");
 }
 
+SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace, void>::
+    SharedAllocationRecord(
+        const Kokkos::Experimental::SYCL& exec_space,
+        const Kokkos::Experimental::SYCLDeviceUSMSpace& space,
+        const std::string& label, const size_t size,
+        const SharedAllocationRecord<void, void>::function_type dealloc)
+    // Pass through allocated [ SharedAllocationHeader , user_memory ]
+    // Pass through deallocation function
+    : base_t(
+#ifdef KOKKOS_ENABLE_DEBUG
+          &SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                                  void>::s_root_record,
+#endif
+          Kokkos::Impl::checked_allocation_with_header(exec_space, space, label,
+                                                       size),
+          sizeof(SharedAllocationHeader) + size, dealloc, label),
+      m_space(space) {
+  SharedAllocationHeader header;
+
+  this->base_t::_fill_host_accessible_header_info(header, label);
+
+  // Copy to device memory
+  Kokkos::Experimental::SYCL exec;
+  Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace, HostSpace>(
+      exec, RecordBase::m_alloc_ptr, &header, sizeof(SharedAllocationHeader));
+  exec.fence(
```

Comment:
I'm still not finished with this pull request but appreciate your feedback already.
I clearly missed adapting this spot to use the provided `exec_space`. We can discuss if this `fence` should be there if an execution space is provided or not since this is still part of the `allocation` which we decided should always be synchronous.
With this pull request, all `View` allocations go through this code path that is called with an execution space argument but it's not difficult to change that.

Overall, this pull request doesn't make a difference for most backends (apart from the fence that we are discussing here). `SYCL` always needs an execution space/queue to do the allocations and so far we were just using the default queue (and this changes with this pull request). `CUDA` only uses a specific stream with `CUDA 11` and another `CMake` argument that is not set by default due to problems with `UCX`.